datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaProducerWrapper.java

```diff
@@ -221,47 +234,63 @@ void send(DatastreamTask task, ProducerRecord<K, V> producerRecord, Callback onC
         }
         // Set a max_send_attempts for KafkaException as it may be non-recoverable
         if (numberOfAttempt > MAX_SEND_ATTEMPTS || ((cause instanceof Error || cause instanceof RuntimeException))) {
-          _log.error("Send failed for partition {} with a non retriable exception", producerRecord.partition(), e);
+          _log.error(String.format("Send failed for partition %d with a non-retriable exception",
+              producerRecord.partition()), e);
           throw generateSendFailure(e);
         } else {
-          _log.warn("Send failed for partition {} with retriable exception, retry {} out of {} in {} ms.",
-              producerRecord.partition(), numberOfAttempt, MAX_SEND_ATTEMPTS, _sendFailureRetryWaitTimeMs, e);
+          _log.warn(String.format(
+              "Send failed for partition %d with a retriable exception, retry %d out of %d in %d ms.",
+              producerRecord.partition(), numberOfAttempt, MAX_SEND_ATTEMPTS, _sendFailureRetryWaitTimeMs), e);
           Thread.sleep(_sendFailureRetryWaitTimeMs);
         }
       } catch (Exception e) {
-        _log.error("Send failed for partition {} with an exception", producerRecord.partition(), e);
+        _log.error(String.format("Send failed for partition %d with an exception", producerRecord.partition()), e);
         throw generateSendFailure(e);
       }
     }
   }
 
-  private synchronized void shutdownProducer() {
-    Producer<K, V> producer = _kafkaProducer;
-    // Nullify first to prevent subsequent send() to use
-    // the current producer which is being shutdown.
-    _kafkaProducer = null;
+  @VisibleForTesting
+  void shutdownProducer() {
+    Producer<K, V> producer;
+    synchronized (_producerLock) {
+      producer = _kafkaProducer;
+      // Nullify first to prevent subsequent send() to use
+      // the current producer which is being shutdown.
+      _kafkaProducer = null;
+    }
+
+    // This may be called from the send callback. The callbacks are called from the sender thread, and must complete
+    // quickly to avoid delaying/blocking the sender thread. Thus schedule the actual producer.close() on a separate
+    // thread
     if (producer != null) {
-      producer.close(TIME_OUT, TimeUnit.MILLISECONDS);
-      NUM_PRODUCERS.decrementAndGet();
+      _producerCloseExecutorService.submit(() -> {
+        _log.info("KafkaProducerWrapper: Closing the Kafka Producer");
+        producer.close(TIME_OUT, TimeUnit.MILLISECONDS);
+        NUM_PRODUCERS.decrementAndGet();
+        _log.info("KafkaProducerWrapper: Kafka Producer is closed");
+      });
     }
   }
 
   private DatastreamRuntimeException generateSendFailure(Exception exception) {
```

Comment:
nit: not sure why this method called `generateSendFailure`. Should this be renamed to `handleSendFailure`?

Reply:
Went back and forth in my head on this. `throws handleSendFailure` looks very weird to me. So leaving this as is.